### ISSUE TITLE
Getting org.eclipse.jetty.io.EofException: Early EOF  after ~90mb files while using SterlingCommerce

### ISSUE THREAD

**Issue by sreedharj** [eclipse/jetty.project#6759]
**Jetty version(s)**   9.4.40

**Java version/vendor** `(use: java -version)`

**OS type/version**

**Description**
When sending  100mb files concurrently  to Jetty Server,, transfer is failing.  With out debug mode 2 concurrent requests are failing.  There are no errors/exceptions logged . With the jetty debug mode,  Early EOF exceptions are seen .  when sending 20 concurrent large files.
[1631191904029]org.eclipse.jetty.io.EofException: Early EOF
        at org.eclipse.jetty.server.HttpInput$3.getError(HttpInput.java:1207)
        at org.eclipse.jetty.server.HttpInput$3.noContent(HttpInput.java:1195)
        at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:333)
        at java.io.InputStream.read(InputStream.java:113)
        at com.sterlingcommerce.woodstock.services.pshttp.servlet.PSHttpServletHandler.streamBodyToDatabase(PSHttpServletHandler.java:684)
        at com.sterlingcommerce.woodstock.services.pshttp.servlet.PSHttpServletHandler.streamDocToDatabase(PSHttpServletHandler.java:508)
        at com.sterlingcommerce.woodstock.services.pshttp.servlet.PSHttpServletHandler.handle(PSHttpServletHandler.java:299)
        at com.sterlingcommerce.woodstock.services.pshttp.servlet.PSHttpServlet.process(PSHttpServlet.java:251)
        at com.sterlingcommerce.woodstock.services.pshttp.servlet.PSHttpServlet.doPost(PSHttpServlet.java:120)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)

have attached the  jetty debug logs and tcpdump.  port is 59113
[tcpdump-jetty-issue.pcap.gz](https://github.com/eclipse/jetty.project/files/7136867/tcpdump-jetty-issue.pcap.gz)

**How to reproduce?**

**Comment by sreedharj** [comment]
[jetty.log.D20210909.zip](https://github.com/eclipse/jetty.project/files/7137124/jetty.log.D20210909.zip)
[jetty.log.D20210909 (2).zip](https://github.com/eclipse/jetty.project/files/7137147/jetty.log.D20210909.2.zip)
[jetty.log.D20210909 (3).zip](https://github.com/eclipse/jetty.project/files/7137150/jetty.log.D20210909.3.zip)
[jetty.log.D20210909 (4).zip](https://github.com/eclipse/jetty.project/files/7137154/jetty.log.D20210909.4.zip)

**Comment by sbordet** [comment]
The only exceptions present in your logs are:

```
java.nio.channels.NotYetConnectedException
	at com.sterlingcommerce.perimeter.api.nio.PsSocketChannel.write(PsSocketChannel.java:385)
	at java.nio.channels.SocketChannel.write(SocketChannel.java:513)
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:273)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:381)
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:823)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:223)
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:550)
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:910)
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:987)
	at org.eclipse.jetty.server.HttpOutput.channelWrite(HttpOutput.java:285)
	at org.eclipse.jetty.server.HttpOutput.complete(HttpOutput.java:515)
	at org.eclipse.jetty.server.Response.completeOutput(Response.java:926)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:538)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$3.run(ChannelEndPoint.java:152)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)
	at java.lang.Thread.run(Thread.java:785)
```

It is very unlikely that `NotYetConnectedException` is thrown for a `SocketChannel` that obviously _was_ connected, because Jetty was able to read a request from it.

Since it is the code at `com.sterlingcommerce.perimeter.api.nio.PsSocketChannel` throwing this weird exception, I would ask/check there -- I don't think this is a Jetty issue at all.

**Comment by gregw** [comment]
I had a look at your trace.   Am I correct in saying that 59113 is the server and 47610 is the client?

Either way, I can see a lot of chunked data being sent from 47610 to 59113 in 2048 byte chunks (you see a 800 on its own line in the data at regular intervals).

There is no terminal chunk of 0 bytes sent, just a final chunk of 800.    So firstly this is a 1 in 2048 chance that the content was exactly a multiple of 800, unlikely.   But a valid message must end in a 0 length chunk to avoid the early EOF exception (assuming this is chunked, which I can't tell for sure without the start of the conversation, but it does look like it).  Also, whatever records your data contains, it looks like they end with a `Z` flag, and the data does not end with such a flag.

All indications are that jetty is correct and that the message was not ended correctly. EarlyEOF is the correct exception to throw.

The close of the TCP connection is initiated by 47610 by sending a FIN.

The trace appears to be almost 3 minutes long (and we don't actually see the start of the request), which is moderately long.   I'm thinking that some intermediary has timed out or reached some other limit.  Note that it really should have RST the connection rather than FIN.

So there is no indication of a jetty problem here.   Also as you have heavily modified jetty with Ps versions of channels and selectors, then I don't think we can deal with this on an open source basis even if there was a suggestion of a jetty issue.   If you need assistence with such heavily modified jetty, then I suggest contacting us about some commercial support.

**Comment by sreedharj** [comment]
thanks @gregw    I still think issue is with the Jetty .  these concurrent large file transfer is working fine with Jetty V 9.4.26,  After upgrading to 9.4.40,  we are  seeing  issue..  In fact,  issue was seen in all the jetty versions   after  9.4.26.  With out jetty debug mode,  2 to 4 concurrent 100m requests itself fails.

**Comment by gregw** [comment]
@sreedharj perhaps the issue is with jetty.  But it is just as likely that jetty stand alone is OK, but perhaps we changed something in the internals that your PS classes are extending and now there is a mismatch.

In the TCP trace, I am clearly seeing an early EOF, as the message is not correctly terminated.   So jetty is correct in throwing an earlyEOF exception.    Now there may be some contributing reason why the client has decided to FIN the connection rather than continue with the message,  but we have no idea what that is.     In order for you to demonstrate that this is a jetty issue, you need to show that either: the message was correctly terminated but jetty said otherwise; or the message was incorrectly terminated because of something jetty did.

I've had a look at the change log of 9.4.27 and nothing jumps out at me as a likely change that could cause your PS integration code problems.   But if you are sure that the issue is in 9.4.26 to 9.4.27, then you could disect that release.   Build locally and bring in one patch at a time until you can reproduce the problem.     Once you identify the specific commit that caused the problem, we would be able to explain what it did and analyse how it may affect you.     Assisting with such disections is something that we can do for our commercial support clients.... but giving some analysis on a disection that you perform is something we can do open source support.

**Comment by joakime** [comment]
@sreedharj 9.4.40 has 3 open security vulnerabilities on it - https://www.eclipse.org/jetty/security_reports.php 
I would not stay with 9.4.40, even if you find out the solution to your sterling issues, have you tried with 9.4.43?

@sreedharj try again without the sterling layer.
If it works without issue, then you know the issue is in that sterling layer.

Doing a quick search online, there's numerous issues with this sterling layer reported on many other open source project issue trackers.

Here's the what's in common with the various solutions to sterling related issues (this information is not official, nor sanctioned by whoever sterling is, it's just gleaned from the successful solutions on other open source issue trackers. we have no experience with sterling, so we cannot verify if these are even valid solutions).
* Always keep sterling up to date, aggressively so.  It should be updated as soon as a new release is made available by whoever manages sterling (most issues with sterling are fixed this way)
* Don't use sterling with non OpenJDK JVMs, such as IBM JVM/JDK, or Android, or other embedded system JVMs (this is the second most common successful fix)
* Don't use sterling with projects built to be run on OpenJDK and it's networking quirks.  Sterling apparently makes bad assumptions that are at odds with how OpenJDK works, especially so with SSL/TLS. (this is the 3rd most common reported solution, don't use sterling with SSL/TLS, or projects that have been updated to support OpenJDK 9 or newer networking behaviors)

**Comment by sreedharj** [comment]
@gregw  @joakime  Thanks for your inputs..  I will try to run these tests with Jetty alone i.e with out PS.   
As I said about same PS code works very well with  SFTP Server, which also uses nio.

**Comment by sreedharj** [comment]
[tcpdump-jetty-issue.zip](https://github.com/eclipse/jetty.project/files/7158969/tcpdump-jetty-issue.zip)
@gregw   I have captured new tcp traces  from the start of the request to end.  This time I have disabled the  Jetty debug logging, so that issue can be seen with 2 concurrent requests itself.

Jetty server port is :  59113

**Comment by gregw** [comment]
@sreedharj your latest trace confirms my previous diagnoisis.   Your stream starts with:
```
POST /hello HTTP/1.1
Transfer-Encoding: chunked
Content-Type: binary/octet-stream
Host: b2bisand2:59113
Connection: Keep-Alive
User-Agent: Apache-HttpClient/4.2.5 (java 1.5)

800
CBNSF838                            CIFO838450                  UYYNNNNNRRDC                               BNSF                                                                CIFP3A                             BNSF3A                                                                                Y
DBTP             000001291                       200304091906    10
DLX              000001
DPLA             4 FC1999072690
DN1              FCCERTEX GULF COAST INC              C50691187220000
DN9              ZAA
DPERX2           ICTE7136743000
DDTM             03699991231
DDTM             30220001108
DN1              FD                                   C50691187220000
DN3              1110 LOCKWOOD DR
DN4              HOUSTON                       TX77020      US COHARRIS
DN1              FE                                   C50691187220000
DN3              PO BOX 18428
DN4              HOUSTON                       TX77223      US
DLX              000002
DPLA             2 FC2003040988
DN1              FCCERTEX GULF COAST INC              C50691187220000
DN9              ZAD
DPERX2           ICTE7136743000
DDTM             03699991231
DDTM             30220030409
DN1              FD                                   C50691187220000
DN3              1110 LOCKWOOD DR
DN4              HOUSTON                       TX77020      US COHARRIS
DN1              FE                                   C50691187220000
DN3              PO BOX 18428
DN4              HOUSTON                       TX77223      US
DERI             C5061329017    01C5228021192    3107
Z

CBNSF838                            CIFO838450                  UYYNNNNNRRDC                               BNSF                                                                CIFP3A                             BNSF3A                                                                                Y
DBTP             000001291                       200304091906    10
DLX              000001
DPLA             4 FC1999072690
DN1              FCCERTEX GULF COAST INC           
800
   C50691187220000
DN9              ZAA
DPERX2           ICTE7136743000
DDTM             03699991231
DDTM             30220001108
DN1              FD                                   C50691187220000
DN3              1110 LOCKWOOD DR
DN4              HOUSTON                       TX77020      US COHARRIS
DN1              FE                                   C50691187220000
DN3              PO BOX 18428
DN4              HOUSTON                       TX77223      US
DLX              000002
DPLA             2 FC2003040988
DN1              FCCERTEX GULF COAST INC              C50691187220000
DN9              ZAD
DPERX2           ICTE7136743000
DDTM             03699991231
DDTM             30220030409
DN1              FD                                   C50691187220000
DN3              1110 LOCKWOOD DR
DN4              HOUSTON                       TX77020      US COHARRIS
DN1              FE                                   C50691187220000
DN3              PO BOX 18428
DN4              HOUSTON                       TX77223      US
DERI             C5061329017    01C5228021192    3107
Z
... etc. etc. etc.
```

We can see it is a chunked POST request.   We can see the first chunk is of size 0x800 (as are all subsequent chunks) and contains some kind of text record that ultimately ends with a single line `Z` and then a blank line.  You can see that the start of the next record is also in the first chunk, and then continues in the next chunk for it ends in `Z` and the pattern continues.

the last chunk sent is:
```
800
50                  UYYNNNNNRRDC                               BNSF                                                                CIFP3A                             BNSF3A                                                                                Y
DBTP             000001291                       200304091906    10
DLX              000001
DPLA             4 FC1999072690
DN1              FCCERTEX GULF COAST INC              C50691187220000
DN9              ZAA
DPERX2           ICTE7136743000
DDTM             03699991231
DDTM             30220001108
DN1              FD                                   C50691187220000
DN3              1110 LOCKWOOD DR
DN4              HOUSTON                       TX77020      US COHARRIS
DN1              FE                                   C50691187220000
DN3              PO BOX 18428
DN4              HOUSTON                       TX77223      US
DLX              000002
DPLA             2 FC2003040988
DN1              FCCERTEX GULF COAST INC              C50691187220000
DN9              ZAD
DPERX2           ICTE7136743000
DDTM             03699991231
DDTM             30220030409
DN1              FD                                   C50691187220000
DN3              1110 LOCKWOOD DR
DN4              HOUSTON                       TX77020      US COHARRIS
DN1              FE                                   C50691187220000
DN3              PO BOX 18428
DN4              HOUSTON                       TX77223      US
DERI             C5061329017    01C5228021192    3107
Z

CBNSF838                            CIFO838450                  UYYNNNNNRRDC                               BNSF                                                                CIFP3A                             BNSF3A                                                                                Y
DBTP             000001291                       200304091906    10
DLX              0
```
We can see this is again a 0x800 sized chunk, so the HTTP message is not correctly terminated with a 0x0 sized chunk.   More over that last chunk finishes one record with `Z` and then starts another that is never terminated.     Thus it is entirely clear that the message is not completely sent and has ended early.  So the EarlyEOF exception is correct.

The question now is why??

things go OK for the first 1.3 seconds of the transfer, then flow control kicks in rate of data sent slows.  No more data is sent (ie the window doesn't open) for over 3 minutes and at 188s the server gives up and sends a '200 OK' response (even though it is clearly not OK) and resets the connection.     This looks to me like at 1.3s the application on the server stopped reading data, so  jetty applied back pressure.  For 3 minutes the application did not read anymore and then something timed out.  The request completed "normally" with an OK, but because the request had not completed OK, the connection was reset instead of being gently closed.

So you need to look at your server application and work out why it stopped reading the data?

**Comment by sreedharj** [comment]
Thanks @gregw  for analyzing the tcp trace.    Time out is happening from the Jetty.  I had configured idle time out as 180 secs.  From the application side,  servlet is waiting for the data . Here is the thread dump.
t java/lang/Object.wait(Native Method) 
at java/lang/Object.wait(Object.java:172(Compiled Code)) 
at org/eclipse/jetty/server/HttpInput.blockForContent(HttpInput.java:584) 
at org/eclipse/jetty/server/HttpInput$1.blockForContent(HttpInput.java:1164) 
at org/eclipse/jetty/server/HttpInput.read(HttpInput.java:330) 
at java/io/InputStream.read(InputStream.java:113(Compiled Code)) 
at com/sterlingcommerce/woodstock/services/pshttp/servlet/PSHttpServletHandler.streamBodyToDatabase(PSHttpServletHandler.java:684) 
at com/sterlingcommerce/woodstock/services/pshttp/servlet/PSHttpServletHandler.streamDocToDatabase(PSHttpServletHandler.java:508) 
at com/sterlingcommerce/woodstock/services/pshttp/servlet/PSHttpServletHandler.handle(PSHttpServletHandler.java:299) 
at com/sterlingcommerce/woodstock/services/pshttp/servlet/PSHttpServlet.process(PSHttpServlet.java:251) 
at com/sterlingcommerce/woodstock/services/pshttp/servlet/PSHttpServlet.doPost(PSHttpServlet.java:120) 
at javax/servlet/http/HttpServlet.service(HttpServlet.java:707) 
at javax/servlet/http/HttpServlet.service(HttpServlet.java:790) 
at org/eclipse/jetty/servlet/ServletHolder.handle(ServletHolder.java:791) 
at org/eclipse/jetty/servlet/ServletHandler.doHandle(ServletHandler.java:550) 
at org/eclipse/jetty/server/handler/ScopedHandler.handle(ScopedHandler.java:143) 
at org/eclipse/jetty/security/SecurityHandler.handle(SecurityHandler.java:602) 
at org/eclipse/jetty/server/handler/HandlerWrapper.handle(HandlerWrapper.java:127)

---
and the selector  thread is waiting on select

at java/lang/Object.wait(Native Method) 
at java/lang/Object.wait(Object.java:172(Compiled Code)) 
at EDU/oswego/cs/dl/util/concurrent/Semaphore.acquire(Bytecode PC:28) 
at EDU/oswego/cs/dl/util/concurrent/SemaphoreControlledChannel.take(Bytecode PC:18) 
at com/sterlingcommerce/perimeter/api/nio/PsSelector.block(PsSelector.java:93) 
at com/sterlingcommerce/perimeter/api/nio/PsSelector.select(PsSelector.java:191) 
at com/sterlingcommerce/perimeter/api/nio/PsSelector.select(PsSelector.java:175) 
at org/eclipse/jetty/io/ManagedSelector.nioSelect(ManagedSelector.java:183(Compiled Code)) 
at org/eclipse/jetty/io/ManagedSelector.select(ManagedSelector.java:190) 
at org/eclipse/jetty/io/ManagedSelector$SelectorProducer.select(ManagedSelector.java:606) 
at org/eclipse/jetty/io/ManagedSelector$SelectorProducer.produce(ManagedSelector.java:543) 
at org/eclipse/jetty/util/thread/strategy/EatWhatYouKill.produceTask(EatWhatYouKill.java:360) 
at org/eclipse/jetty/util/thread/strategy/EatWhatYouKill.doProduce(EatWhatYouKill.java:184(Compiled Code)) 
at org/eclipse/jetty/util/thread/strategy/EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) 
at org/eclipse/jetty/util/thread/strategy/EatWhatYouKill.run(EatWhatYouKill.java:129) 
at org/eclipse/jetty/util/thread/ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:383) 
at org/eclipse/jetty/util/thread/QueuedThreadPool.runJob(QueuedThreadPool.java:882) 
at org/eclipse/jetty/util/thread/QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036) 
at java/lang/Thread.run(Thread.java:785)

---
even though data is available for read,  somehow selector is not getting notified.

**Comment by gregw** [comment]
Then it indeed looks like a problem with the PsSelector code.   Can you tell us more about what it is and why you are using it?

**Comment by sreedharj** [comment]
@gregw  I think, some race condition in the jetty code.  I have added few debug statements in the PS code.  I see,  read calls from Jetty thread, even when the  read is not ready.

ava.lang.Thread.getStackTrace(Thread.java:1117)
 com.sterlingcommerce.perimeter.api.nio.PsSocketChannel.read(PsSocketChannel.java:261)
 com.sterlingcommerce.perimeter.api.nio.PsSocketChannel.read(PsSocketChannel.java:242)
 org.eclipse.jetty.io.ChannelEndPoint.fill(ChannelEndPoint.java:246)
 org.eclipse.jetty.server.HttpConnection.fillRequestBuffer(HttpConnection.java:345)
 org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
 org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
 org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
 org.eclipse.jetty.io.ChannelEndPoint$3.run(ChannelEndPoint.java:152)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
 org.eclipse.jetty.io.ManagedSelector$$Lambda$222.00000000738FEB40.run(Unknown Source)
 org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)
 org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)
 java.lang.Thread.run(Thread.java:785)]

and when the  read Ready is set, there is no interest registered from the Jetty.

In the above stack trace,   org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105),  fillable is called when  read ready right?

**Comment by sreedharj** [comment]
A perimeter server is a software tool for communications management that can be installed in a DMZ. The perimeter server manages the communications flow between outer layers of your network and the TCP-based transport adapters. A perimeter server can solve problems with network congestion, security, and scalability, especially in high-volume, Internet-gateway environments. .   https://www.ibm.com/docs/en/b2b-integrator/6.0.0?topic=integrator-perimeter-servers

**Comment by gregw** [comment]
@sreedharj,

It is perfectly reasonable to try reading before registering interest.   Using a selector to register interest is a moderately expensive operation, thus jetty will always attempt a read first and only register with the selector for read interest if the read returned 0 bytes read.

But in the stack trace you have provided, there is interest registered because there is a callback available to the FillInterest object.  So in this case it looks like a normal selector wakeup followed by a read.  I see no evidence of a race condition in jetty.

Have you raised this issue with sterlingcommerce ?

**Comment by sreedharj** [comment]
@gregw   I am the  part of Sterling Commerce team  and trying to troubleshoot this issue .   

I have added few debug statements into the PS code  , interest register, select, read operations went fine for some time,  after read call from Jetty, there is interest registered. 

[2021-09-15 11:11:49.254] WARN HttpServerAdapter_HttpServerAdapter_node1-Thread-12 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 0
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKey.interestOps(i) called on [hash=[846b87b9], valid=true, channel=[49948f0d,lrdy=5], selector=[3ad6f8ac], interestBits=0] with ops=1, old=0, ready=5
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKey.interestOps(i) called on [hash=[846b87b9], valid=true, channel=[49948f0d,lrdy=5], selector=[3ad6f8ac], interestBits=0] with ops=1 [java.lang.Thread.getStackTraceImpl(Native Method)
 java.lang.Thread.getStackTrace(Thread.java:1117)
 com.sterlingcommerce.perimeter.api.nio.PsSelectionKey.interestOps(PsSelectionKey.java:131)
 org.eclipse.jetty.io.ChannelEndPoint.updateKey(ChannelEndPoint.java:378)
 org.eclipse.jetty.io.ChannelEndPoint.updateKeyAction(ChannelEndPoint.java:357)
 org.eclipse.jetty.io.ChannelEndPoint$$Lambda$224.00000000A4008C30.update(Unknown Source)
 org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processUpdates(ManagedSelector.java:568)
 org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:539)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
 org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:383)
 org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)
 org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)
 java.lang.Thread.run(Thread.java:785)]
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 1


[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] after ReadData: 8192 liveReady: 4
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] existing: 0liveReady: 4

[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] existing: 0liveReady: 4

[2021-09-15 11:11:49.254] WARN HttpServerAdapter_HttpServerAdapter_node1-Thread-12 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 0

[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 1
[2021-09-15 11:11:49.254] WARN #### PsSelectionKey.interestOps(i) adding selected
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 1
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKeyhash=[4bff1069], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 1
[2021-09-15 11:11:49.254] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKey.interestOps(i) called on [hash=[846b87b9], valid=true, channel=[49948f0d,lrdy=5], selector=[3ad6f8ac], interestBits=1] with ops=0, old=1, ready=5

[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKey.interestOps(i) called on [hash=[846b87b9], valid=true, channel=[49948f0d,lrdy=5], selector=[3ad6f8ac], interestBits=1] with ops=0 [java.lang.Thread.getStackTraceImpl(Native Method)
 java.lang.Thread.getStackTrace(Thread.java:1117)
 com.sterlingcommerce.perimeter.api.nio.PsSelectionKey.interestOps(PsSelectionKey.java:131)
 org.eclipse.jetty.io.ChannelEndPoint.updateKey(ChannelEndPoint.java:378)
 org.eclipse.jetty.io.ManagedSelector$SelectorProducer.updateKeys(ManagedSelector.java:714)
 org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:541)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
 org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:383)
 org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)
 org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)
 java.lang.Thread.run(Thread.java:785)]
[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-621 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 0

[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[5], state=[CONNECTED] before ReadData: liveReady: 5
[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### **PsSocketChannel.turnOffLiveReadyBits(i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] turned off 1, new live=4
[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] after ReadData: 8192 liveReady: 4**

[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] existing: 0liveReady: 4
**[2021-09-15 11:11:49.255] WARN qtp_HttpServerAdapter_HttpServerAdapter_node1-617 #### PsSocketChannel.read(BB[],i,i) hash=[49948f0d], liveReady=[4], state=[CONNECTED] existing: 0liveReady: 4**

[2021-09-15 11:11:49.255] WARN HttpServerAdapter_HttpServerAdapter_node1-Thread-17 #### PsSelectionKeyhash=[49948f0d], liveReady=[5], state=[CONNECTED] liveReady: 5 interestBits: 0

thread name starting with qtp_  is  Jetty threads,      and  thread name starting with HTTPServerAdapter is PS thread.   In the above qtp_HttpServerAdapter_HttpServerAdapter_node1-617  made read call on the PsSocketchannel, read 8192 bytes and then called again read 2 times.  After that it did not call interest register.

**Comment by sreedharj** [comment]
@gregw   @joakime  

[jetty.log.D20210915.zip](https://github.com/eclipse/jetty.project/files/7172483/jetty.log.D20210915.zip)
 I have captured one more jetty log with  org.eclipse.jetty.io.LEVEL=DEBUG,     

[2021-09-15 23:32:01.443] ALL 000000000000 GLOBAL_SCOPE 2021-09-15 23:32:01.443:DBUG:oeji.ChannelEndPoint:qtp_HttpServerAdapter_HttpServerAdapter_node1-640: filled 0 HeapByteBuffer@32ed86f7[p=0,l=0,c=8192,r=0]={<<<>>>         ...  FE   }
[2021-09-15 23:32:01.443] ALL 000000000000 GLOBAL_SCOPE 2021-09-15 23:32:01.443:DBUG:oeji.ChannelEndPoint:qtp_HttpServerAdapter_HttpServerAdapter_node1-640: filled 0 HeapByteBuffer@32ed86f7[p=0,l=0,c=8192,r=0]={<<<>>>         ...  FE   }
[2021-09-15 23:32:01.443] ALL 000000000000 GLOBAL_SCOPE 2021-09-15 23:32:01.443:DBUG:oeji.FillInterest:qtp_HttpServerAdapter_HttpServerAdapter_node1-640: interested FillInterest@5254317e{org.eclipse.jetty.server.HttpConnection$BlockingReadCallback@184bea50}
[2021-09-15 23:32:01.443] ALL 000000000000 GLOBAL_SCOPE 2021-09-15 23:32:01.443:DBUG:oeji.ChannelEndPoint:qtp_HttpServerAdapter_HttpServerAdapter_node1-640: changeInterests p=false 0->1 for SocketChannelEndPoint@f5dffb0c{l=/10.15.102.66:59113,r=/10.15.105.122:48642,OPEN,fill=FI,flush=-,to=0/180000}{io=0/1,kio=0,kro=4}->HttpConnection@b7314128[p=HttpParser{s=CHUNK,58630144 of -1},g=HttpGenerator@c899b59{s=START}]=>HttpChannelOverHttp@f0e825c2{s=HttpChannelState@c0d620a6{s=HANDLING rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=1,c=false/false,a=HANDLING,uri=//b2bisand2:59113/hello,age=18503}
[2021-09-15 23:32:01.443] ALL 000000000000 GLOBAL_SCOPE 2021-09-15 23:32:01.443:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-640: Queued change lazy=false org.eclipse.jetty.io.ChannelEndPoint$$Lambda$213/0000000034007FE0@3bb6bd55 on ManagedSelector@5f2852ab{STARTED} id=0 keys=3 selected=0 updates=1

So not sure, if the read interest was registered  after the last read .  

In the attached log  you can see the above log statements.  After this timestamp :  2021-09-15 23:32:01.443,,    IDLE time out occurred after 180 secs.

**Comment by gregw** [comment]
@sreedharj are you able to try jetty-10?   The HttpInput of Jetty-10 has been significantly reworked and simplified and is a lot easier to debug as a result.    
The other option is for you to bisect the changes between 9.4.26 to 9.4.27 to identify the exact change in behaviour that is causing your problem.

**Comment by gregw** [comment]
@sreedharj another idea, can you run the jetty-io tests from a checkout of jetty using your Ps layer.   Hopefully this may reproduce your problem and give a simpler reproduction on which to diagnose.

**Comment by sreedharj** [comment]
@gregw   does the above jetty log gives any clue on the  read interest registration?   Looks like it has updated the key with the interest as 0 and after that nothing happened

I can try with jetty-10.  I hope, there won't be any integration issues  with webapp deployment etc.

I tried building  jetty-io  but was not successful , as it getting time out during the tests execution, so built the project by skipping the tests.   is there a way to increase the time out for the test exeuction

**Comment by gregw** [comment]
@sreedharj the log `changeInterests p=false 0->1` says that the interest ops where changed from 0 to 1, ie changed to being read interested.   But note that this is our version of the interest ops and is not immediately set on the selector.   Since changing interest is moderately expensive and it is a frequent change to see interest go 1->0->1 we only do the actual update just before the selection.  So if we think the interest was 1 at the last selection, we change out desired ops to 0 when it is selected and this log shows it changing back to 1.  

So I can see `io=0/1` on the EndPoint which means the current interest ops are 0, but the desired are 1.  Thus you should see a message like: `changeInterests p=true 0->1` when the actual key is updated.   In the debug you listed I can see that change queued, but you've not given enough debug to see it actually executed.

What test times out for you?  It takes my little laptop only 1:30 to build and test jetty-io, so most machines should be able to do this without bother.  You can use the normal maven flags to build without tests: `mvn -DfailIfNoTests=false -Dtest=none install`.   But ultimately we want to run with tests.    How about just loading into your IDE and running the jetty-io tests there?

**Comment by sreedharj** [comment]
@gregw   please let me know which package you want to me enable the debug to see the queued execution details.

on the jetty-io, I was able to run the tests now.  It was failing because of my code changes in the ManagedSelector class, , after reverting back ,  it was working.

So for now, lets focus on 9.4.40  only,  based on your above log analysis, we just need to figure out,  queued interest change was actually executed or not?  let me know, which package to enable for the debug, will run the test again

In parallel,  I am trying to upgrade the jetty to 10.0.6,

**Comment by sreedharj** [comment]
@gregw   I have added debug statement  ManagedSelector#submit() method. as below 

    private void submit(SelectorUpdate update, boolean lazy)
    {
        if (LOG.isDebugEnabled())
            LOG.debug("Queued change lazy={} {} on {}", lazy, update, this);

        Selector selector = null;
        synchronized (ManagedSelector.this)
        {
            _updates.offer(update);

            if (_selecting && !lazy)
            {
                selector = _selector;
                // To avoid the extra select wakeup.
                _selecting = false;
            }
        }

        if (selector != null)
        {
            if (LOG.isDebugEnabled())
                LOG.debug("Wakeup on submit {}", this);
            selector.wakeup();
        } 
        else 
        {
            if (LOG.isDebugEnabled())
                LOG.debug("selector is null {}",  this);
        }
    }

After running the test,   I see , selector was null, that's the reason , there is no updates to the key.

[2021-09-16 11:59:40.557] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.557:DBUG:oeji.ChannelEndPoint:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: filled 0 HeapByteBuffer@d5845283[p=0,l=0,c=8192,r=0]={<<<>>>\r\n800\r\n  ...RX2    }
[2021-09-16 11:59:40.557] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.557:DBUG:oeji.ChannelEndPoint:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: filled 0 HeapByteBuffer@d5845283[p=0,l=0,c=8192,r=0]={<<<>>>\r\n800\r\n  ...RX2    }
[2021-09-16 11:59:40.557] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.557:DBUG:oeji.FillInterest:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: interested FillInterest@e49c8914{org.eclipse.jetty.server.HttpConnection$BlockingReadCallback@c435aa6a}
[2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.557:DBUG:oeji.ChannelEndPoint:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: changeInterests p=false 0->1 for SocketChannelEndPoint@210fec41{l=/10.15.102.66:59113,r=/10.15.105.122:56450,OPEN,fill=FI,flush=-,to=0/180000}{io=0/1,kio=0,kro=4}->HttpConnection@a37ba572[p=HttpParser{s=CHUNK,1611748 of -1},g=HttpGenerator@3098f44f{s=START}]=>HttpChannelOverHttp@f2443b8c{s=HttpChannelState@f2b6b682{s=HANDLING rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=1,c=false/false,a=HANDLING,uri=//b2bisand2:59113/hello,age=1900}
[2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.558:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: Queued change lazy=false org.eclipse.jetty.io.ChannelEndPoint$$Lambda$215/000000008400B000@5013b1e8 on ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=1
[2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.558:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: **selector is null** ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=2

selector was null, because  _selecting  param is set to false.     

So we need to figure out  why _selecting is set to false, before this submit call?

**Comment by gregw** [comment]
typically _selecting is set to false when a previous submit has already done a wakeup on teh current cycle, so there is no need for another wakeup call (which is expensive) as the selector is already going to wakeup.

Does the selector wakeup after this?

**Comment by sreedharj** [comment]
@gregw  no,  selector wakeup  did not happen.  Please see the attached log
[jetty.log.D20210916.zip](https://github.com/eclipse/jetty.project/files/7176354/jetty.log.D20210916.zip)

**Comment by sreedharj** [comment]
just waited for the Idle time out to occur

2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.558:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: Queued change lazy=false org.eclipse.jetty.io.ChannelEndPoint$$Lambda$215/000000008400B000@5013b1e8 on ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=1
[2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.558:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: selector is null ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=2
[2021-09-16 12:02:38.563] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 12:02:38.561:DBUG:oeji.IdleTimeout:Connector-Scheduler-6c4c3238-1: SocketChannelEndPoint@7e65b823{l=/10.15.102.66:59113,r=/10.15.105.122:56448,OPEN,fill=FI,flush=-,to=178006/180000}{io=0/1,kio=0,kro=5}->HttpConnection@5f3c3242[p=HttpParser{s=CHUNK,1559668 of -1},g=HttpGenerator@c1b373d4{s=START}]=>HttpChannelOverHttp@ea887a21{s=HttpChannelState@1b113465{s=HANDLING rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=1,c=false/false,a=HANDLING,uri=//b2bisand2:59113/hello,age=179911} idle timeout check, elapsed: 178005 ms, remaining: 1995 ms

**Comment by sreedharj** [comment]
@gregw 
[2021-09-16 11:59:40.556] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.556:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-623: Queued change lazy=false org.eclipse.jetty.io.ChannelEndPoint$$Lambda$215/000000008400B000@57f4ac62 on ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=0
[2021-09-16 11:59:40.556] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.556:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-623: Wakeup on submit ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=1

in the above  interest change was queued up and then immediately wakeup was executed.


[2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.558:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: Queued change lazy=false org.eclipse.jetty.io.ChannelEndPoint$$Lambda$215/000000008400B000@5013b1e8 on ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=1
[2021-09-16 11:59:40.558] ALL 000000000000 GLOBAL_SCOPE 2021-09-16 11:59:40.558:DBUG:oeji.ManagedSelector:qtp_HttpServerAdapter_HttpServerAdapter_node1-621: selector is null ManagedSelector@f75d8c14{STARTED} id=0 keys=3 selected=0 updates=2

Here  interest changed queued up but there is no wakup  call,  it is already executed at 11:59:40.556.    

So somehow  this _selecting was not getting updated to true.

**Comment by gregw** [comment]
So how sure are that the PsSelector is handling calls to wakeup correctly?

We set _selecting to be true just before entering selecting.  We set it to false after we have called wakeup.      I believe the normal selector deals correctly with the small window that exists there that might mean we call wakeup just before select is called, and the normal selector does wakeup.    

Perhaps your PsSelector is not handling this condition and if wakeup is called just before select is called, then it doesn't wake up?

I'm only 90% sure the above description is correct as I've not had the time to look at the code in detail (and probably wont until next week).  But it is plausible that the race you are dealing with is in this area. You need to work out why it is not a problem for the normal JVM selector, but is a problem with yours.

It would be really good if you could take one of our test and create a reproduceable test case that you can give us to analyse.

**Comment by sreedharj** [comment]
@gregw  let me add  few debug statements in PsSelector wakeup method

**Comment by sreedharj** [comment]
@gregw  I have done further analysis on the PsSelector wakeup method.    Some times, this wakeup method is not able to  notify the selector, as it is trying to  process the selected keys and return. During this time wakeup method will not be able to notify the selector as it is not able to get the permit ( semaphore)..  So for now, I have added retry logic for the wakeup call to wait for  1ms to notify, if not retry for 20  times.   With this retry logic in place,  i am able to process the large file concurrently. 

There are few times, where wakeup() call was not successful even after 20 ms ( 20 times retry ), but did not see the previous behavior ( where application was not reading the data )..

So, on the jetty side, where updates are queued up for   selector updates,  can we  not trigger one more wake up call,  when queued up updates are more than  certain no?

**Comment by gregw** [comment]
@sreedharj thanks for the update.  It is what we suspected about `PsSelector.wakeup()`! 

Reducing the number of wakeup calls is an important optimization in jetty.   We certainly don't want to do more frequent wakeups in normal jetty.

Potentially we could change the `submit` and `wakeup` methods from private to protected, so that you could override them with implementations that always did a wakeup on a submit and a wakeup always really did a real wakeup on the selector.     Would that be sufficient for your usage? ie can you extend the ManagedSelector in your setup?

But ultimately, I think the best fix is for you to find someway to fix the `PsSelector.wakeup()` call.      That method returns no indication that can be used to know if it succeeded or failed to do the wakeup, so unless it always does the wakeup, you will be committing the code above to make multiple wakeup calls that are at best wasted effort and inefficient. At worst they still might not be enough to actually wakeup the selector.    So you will end up with a slow server that is still not guaranteed to always work!

If you look at the epoll implementation of wakeup, it achieves a wakeup by having a `fd1` file descriptor and `wakeup` is essentially just writing a 0 byte to that FD, so the selector selects it.   Surely you could either let that call through without first grabbing your semaphore, or implement something similar with another special loopback registration you can write to.

I'm going to close this issue now, as it is not a jetty bug.   If Sterling Commerce would like to wrok closely with us to get a highly performant and correct solution, then please do contact us directly to consider commercial support.

**Comment by sreedharj** [comment]
Thanks @gregw .  if we can optimize the jetty code to reduce the wakeup calls on the selector, then that would really improve the performance, may not get into this situation.   Essentially, one jetty thread is updating the key and another jetty thread is calling wakeup on the same selector., 

com.sterlingcommerce.perimeter.api.nio.PsSelector.softWakeup(PsSelector.java:220)
 com.sterlingcommerce.perimeter.api.nio.PsSelector.addSelected(PsSelector.java:68)
 com.sterlingcommerce.perimeter.api.nio.PsSelectionKey.interestOps(PsSelectionKey.java:143)
 org.eclipse.jetty.io.ChannelEndPoint.updateKey(ChannelEndPoint.java:378)
 org.eclipse.jetty.io.ChannelEndPoint.updateKeyAction(ChannelEndPoint.java:357)
 org.eclipse.jetty.io.ChannelEndPoint$$Lambda$217.0000000098007BE0.update(Unknown Source)
 org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processUpdates(ManagedSelector.java:568)
 org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:539)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
 org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
 org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:383)
 org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)
 org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)
 java.lang.Thread.run(Thread.java:785)]

not sure, if overriding the ManagedSelector methods submit and wakeup are going to be helpful, as the issue seems to be  with concurrent calls   updatekeys and wakeup calls from jetty threads on the same selector.
